### DISCUSSION TITLE
Рекомендации с LastFM

### DISCUSSION THREAD

**Discussion by ScreNN** [Chimildic/goofy#205] [Помощь с алгоритмом] [answered]
Привет.

Я может что-то упустил, но не могу сообразить скрипт, чтобы допустим раз в неделю вытягивать рекомендации с LastFM в Спотифай. Вот тут есть пример [https://github.com/Chimildic/goofy/discussions/91](https://github.com/Chimildic/goofy/discussions/91), но там про какое то прошлое речь :)

```
function getRecLastFM() {
let tracks = Lastfm.getRecomStation('Login', 2);

    Playlist.saveWithReplace({
        // id: 'LastFM',
        name: 'LastFM',
        tracks: tracks,
        randomCover: 'update',
        public: false,
    });
}
```
Так что ли?
Ну, треки он вытянул, но как то криво. Вытянул не все рекомендации, а некоторые рекомендации вообще неправильные. То есть, исполнитель правильный, но название трека вообще другое, не такое как в LastFM. Может надо еще и API ласта вставить в config? Или он нужен только для того чтобы отслеживать прослушанное? А без API, как он увидит рекомендации мои?

И тут же вопрос про файл config.gs
```
    UserProperties.setProperty('LASTFM_API_KEY', 'вашеЗначение');
    UserProperties.setProperty('LASTFM_LOGIN', 'вашЛогин');
    UserProperties.setProperty('LASTFM_RANGE_RECENT_TRACKS', '30');
```

Эти строки нужны только для того чтобы Goofy смотрел историю прослушиваний на LastFM? Или их как то еще юзать можно с LastFM

**Comment by Chimildic** [comment, marked as answer]
`LASTFM_API_KEY` нужен для любых запросов к ластфм. Логин понадобится для запроса истории и в некоторых других случаях в качестве дефолт-значения. Например, чтобы не указывать его при вызове [getCustomTop](https://chimildic.github.io/goofy/#/reference/lastfm?id=getcustomtop).

> треки он вытянул, но как то криво

С чем вы сравниваете? Функция [getRecomStation](https://chimildic.github.io/goofy/#/reference/lastfm?id=getrecomstation) просто парсит ответ от радиостанции ластфм указанного пользователя. Она дает 20-30 треков на один запрос. Качество рекомендаций на совести ластфм.

Это радио находится в правом блоке на вашей странице с общей статистикой. 

![image](https://user-images.githubusercontent.com/40138097/187886269-c992bdf8-9257-405f-9f3a-aea1678b8095.png)

Возможно, искомого трека нет в базе Spotify, поэтому подмешиваются аналоги или другие треки исполнителя. Но с [версии 1.8.0](https://chimildic.github.io/goofy/#/changelog?id=Версия-180) алгоритм совпадения был изменен и совсем левых треков должно быть существенно меньше. Попробуйте поднять значение [параметра MIN_DICE_RATING ](https://chimildic.github.io/goofy/#/config?id=Общее).

Если вы говорите о странице _music > your recommendations_, то ее нельзя парсить и нет API-функции для этого. Оно и не нужно. Потому что это просто собранные в кучу похожие элементы на основе истории. Можно по своему локализовать объем таких элементов через отбор входной выборки. Например, похожие треки из топа за неделю.
```js
function updateWeekRecomPlaylist() {
  let topWeekTracks = Lastfm.getTopTracks({
    user: KeyValue.LASTFM_LOGIN,
    period: '7day',
    limit: 10,
  })

  let recomTracks = Lastfm.getSimilarTracks(topWeekTracks, 0.7, 4)
  Combiner.push(recomTracks, Source.craftTracks(topWeekTracks))
  
  Filter.removeTracks(recomTracks, RecentTracks.get(5000))
  Order.sort(recomTracks, 'meta.popularity', 'desc')

  Playlist.saveWithReplace({
    name: 'Рекомендации по недельному топу',
    tracks: Selector.sliceFirst(recomTracks, 40)
  })
}
```

**Comment by ScreNN** [comment]
Понял. Спасибо.
